lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -347,6 +347,49 @@ SILInstruction *TempRValueOptPass::getLastUseWhileSourceIsNotModified(
   return nullptr;
 }
 
+/// Tries to move an end_access down to extend the access scope over all uses
+/// of the temporary. For example:
+///
+///   %a = begin_access %src
+///   copy_addr %a to [initialization] %temp : $*T
+///   end_access %a
+///   use %temp
+///
+/// We must not replace %temp with %a after the end_access. Instead we try to
+/// move the end_access after "use %temp".
+bool TempRValueOptPass::extendAccessScopes(
+    CopyAddrInst *copyInst, SILInstruction *lastUseInst) {
+
+  SILValue copySrc = copyInst->getSrc();
+  EndAccessInst *endAccessToMove = nullptr;
+  auto begin = std::next(copyInst->getIterator());
+  auto end = std::next(lastUseInst->getIterator());
+
+  for (SILInstruction &inst : make_range(begin, end)) {
+    if (auto *endAccess = dyn_cast<EndAccessInst>(&inst)) {
+      // Is this the end of an access scope of the copy-source?
+      if (!aa->isNoAlias(copySrc, endAccess->getSource())) {
+        // To keep things simple, we can just move a single end_access.
+        if (endAccessToMove)
+          return false;
+        endAccessToMove = endAccess;
+      }
+    } else if (endAccessToMove) {
+      // We cannot move an end_access over a begin_access. This would destroy
+      // the proper nesting of accesses.
+      if (isa<BeginAccessInst>(&inst))
```

Comment:
@eeckstein We also still have an "unpaired" begin access, so strictly speaking, finding all the access scope beginnings means doing this:
`isa<BeginAccessInst>(inst) || isa<BeginUnpairedAccessInst>(inst)`
I'm surprised there's in no helper for it yet.

Reply:
The begin_unpaired_access is only used in the KeyPath runtime functions. And we compile the stdlib without access enforcement. So it should not be a problem.
But checking for the unpaired access does not harm. I'll add it.